File: javascript/manga/src/en/weebcentral.js
const mangayomiSources = [{
    "id": 693275080,
    "name": "Weeb Central",
    "lang": "en",
    "baseUrl": "https://weebcentral.com",
    "apiUrl": "",
    "iconUrl": "https://www.google.com/s2/favicons?sz=128&domain=https://weebcentral.com",
    "typeSource": "single",
    "itemType": 0,
    "version": "0.1.5",
    "pkgPath": "manga/src/en/weebcentral.js"
}];

class DefaultExtension extends MProvider {
    constructor() {
        super();
        this.client = new Client();
    }

    // ── DYNAMIC HEADERS ──────────────────────────────────────────────────────
    // This allows the code to work on iOS, Android, and PC by adapting
   getHeaders(url) {
    const isImage = url.includes("compsci88.com") || url.includes(".webp");
    
    // This pulls the EXACT string your app is currently using in Settings
    const appUserAgent = this.client.userAgent || "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

    return {
        "Referer": "https://weebcentral.com/",
        "User-Agent": appUserAgent,
        "Accept": isImage ? "image/avif,image/webp,*/*" : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        "Sec-Fetch-Dest": isImage ? "image" : "document",
        "Sec-Fetch-Mode": isImage ? "no-cors" : "navigate",
        "Sec-Fetch-Site": isImage ? "cross-site" : "same-origin"
    };
}

    async request(slug, isData = false) {
        const url = `${this.source.baseUrl}${slug}`;
        const headers = {
            "Referer": "https://weebcentral.com/",
            "User-Agent": this.client.userAgent || "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
            "Accept": isData ? "application/json, text/plain, */*" : "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "X-Requested-With": "XMLHttpRequest", // Forces bypass on some CF configs
            "Sec-Fetch-Mode": "cors",
            "Sec-Fetch-Site": "same-origin"
        };

        const res = await this.client.get(url, headers);
        
        if (res.statusCode === 403) {
            throw new Error("Cloudflare Level 2 Block: Please Login in WebView to sync account cookies.");
        }
        return new Document(res.body);
    }

    async getDetail(url) {
        // Use the 'isData' flag for the chapter list call
        const slug = url.includes("weebcentral.com") ? url.split("weebcentral.com")[1] : url;
        const doc = await this.request(slug, false); // Get HTML
        
        // When fetching the chapter list, it's a data call
        const chapSlug = `${slug}/full-chapter-list`;
        const chapDoc = await this.request(chapSlug, true); 
        
        // ... rest of your parsing logic
    }
    // ── NAVIGATION ───────────────────────────────────────────────────────────

    async getPopular(page) {
        const filters = this.getFilterList();
        filters[0].state = 2; // Popularity
        return await this.search("", page, filters);
    }

    async getLatestUpdates(page) {
        const filters = this.getFilterList();
        filters[0].state = 5; // Latest
        return await this.search("", page, filters);
    }

    getImageUrl(id) { 
        return `https://temp.compsci88.com/cover/normal/${id}.webp`; 
    }

    async search(query, page, filters) {
        if (!filters || !filters[0]) filters = this.getFilterList();

        const offset = 32 * (parseInt(page) - 1);
        const sort = filters[0].values[filters[0].state].value;
        const order = filters[1].values[filters[1].state].value;
        const official = filters[2].values[filters[2].state].value;

        // Build the query string manually to ensure accuracy
        const slug = `/search/data?limit=32&offset=${offset}&author=&text=${encodeURIComponent(query)}&sort=${sort}&order=${order}&official=${official}&display_mode=Full%20Display`;
        
        const doc = await this.request(slug);
        const list = [];
        const mangaElements = doc.select("article:has(section)");
        
        for (const manga of mangaElements) {
            const img = manga.selectFirst("img");
            const details = manga.selectFirst("section > a");
            const titleElement = manga.selectFirst("article > div > div > div");
            
            if (details && titleElement) {
                list.push({
                    name: titleElement.text.trim(),
                    imageUrl: img ? img.getSrc : "",
                    link: details.getHref
                });
            }
        }

        const hasNextPage = doc.select("button").length > 0;
        return { list, hasNextPage };
    }

    // ── DETAILS & CHAPTERS ───────────────────────────────────────────────────

    async getDetail(url) {
        const slug = url.includes("weebcentral.com") ? url.split("weebcentral.com")[1] : url;
        const doc = await this.request(slug);
        
        const description = doc.selectFirst("p.whitespace-pre-wrap.break-words")?.text || "";
        const id = slug.split('/').filter(Boolean).pop();
        const imageUrl = this.getImageUrl(id);

        const genres = [];
        let author = "Unknown";
        let status = 5;

        const infoItems = doc.select("ul.flex.flex-col.gap-4 > li");
        for (const li of infoItems) {
            const text = li.text;
            if (text.includes("Author(s):")) author = li.selectFirst("a")?.text || author;
            if (text.includes("Tags(s):")) li.select("a").forEach(a => genres.push(a.text));
            if (text.includes("Status:")) status = this.statusCode(li.selectFirst("a")?.text);
        }

        const chapDoc = await this.request(`${slug}/full-chapter-list`);
        const chapters = chapDoc.select("div.flex.items-center").map(chap => {
            const name = chap.selectFirst("span.grow.flex.items-center.gap-2")?.selectFirst("span")?.text || "Unknown Chapter";
            const dateStr = chap.selectFirst("time.text-datetime")?.text;
            const chUrl = chap.selectFirst("input")?.attr("value");
            
            return {
                name: name,
                url: chUrl,
                dateUpload: dateStr ? String(new Date(dateStr).getTime()) : null
            };
        });

        return { description, imageUrl, author, genre: genres, status, chapters };
    }

   async getPageList(id) {
        const slug = `/chapters/${id}/images?current_page=1&reading_style=long_strip`;
        const doc = await this.request(slug);
        const images = doc.select("section > img");

        return images.map(img => {
            const src = img.attr("src");
            // Extract the domain (e.g., temp.compsci88.com)
            const host = src.split('//')[1].split('/')[0];

            return {
                url: src,
                headers: {
                    "Referer": `${this.source.baseUrl}/`,
                    "User-Agent": this.client.userAgent,
                    "Host": host, // CRITICAL: Must match the image URL domain
                    "Accept": "image/avif,image/webp,*/*",
                    "Sec-Fetch-Dest": "image",
                    "Sec-Fetch-Mode": "no-cors",
                    "Sec-Fetch-Site": "cross-site"
                }
            };
        });
    

        return images.map(img => {
            const src = img.attr("src");
            return {
                url: src,
                headers: {
                    "Referer": "https://weebcentral.com/",
                    "User-Agent": this.client.userAgent,
                    "Accept": "image/avif,image/webp,image/apng,*/*;q=0.8"
                }
            };
        });
    }

    statusCode(status) {
        return { "Ongoing": 0, "Complete": 1, "Hiatus": 2, "Canceled": 3 }[status] ?? 5;
    }

    getFilterList() {
        // ... (Keep your existing filter list here)
        return [
            {
                type_name: "SelectFilter",
                name: "Sort",
                state: 0,
                values: [
                    { type_name: 'SelectOption', name: "Best Match", value: "Best Match" },
                    { type_name: 'SelectOption', name: "Popularity", value: "Popularity" },
                    { type_name: 'SelectOption', name: "Latest Updates", value: "Latest Updates" }
                ]
            },
            {
                type_name: "SelectFilter",
                name: "Order",
                state: 1,
                values: [
                    { type_name: 'SelectOption', name: "Ascending", value: "Ascending" },
                    { type_name: 'SelectOption', name: "Descending", value: "Descending" }
                ]
            },
            {
                type_name: "SelectFilter",
                name: "Official Translation",
                state: 0,
                values: [
                    { type_name: 'SelectOption', name: "Any", value: "Any" },
                    { type_name: 'SelectOption', name: "True", value: "True" },
                    { type_name: 'SelectOption', name: "False", value: "False" }
                ]
            }
            // (Shortened for brevity, keep your full lists for Status, Type, and Tags)
        ];
    }
}

const extension = new DefaultExtension();

File: dart/manga/src/en/weebcentral/weebcentral.dart
import 'package:mangayomi/bridge_lib.dart';
import 'dart:convert';

class WeebCentral extends MProvider {
  WeebCentral({required this.source});

  MSource source;

  final Client client = Client(source);

  @override
  Future<MPages> getPopular(int page) async {
    final res = (await client.get(
      Uri.parse("${source.baseUrl}/popular?page=$page"),
      headers: getHeader(source.baseUrl),
    )).body;
    return parseMangaList(res);
  }

  @override
  Future<MPages> getLatestUpdates(int page) async {
    final res = (await client.get(
      Uri.parse("${source.baseUrl}/latest?page=$page"),
      headers: getHeader(source.baseUrl),
    )).body;
    return parseMangaList(res);
  }

  // Common parser for manga list pages
  MPages parseMangaList(String html) {
    List<MManga> mangaList = [];
    final mangas = parseHtml(html).select("div.manga-list > div.manga-item");
    for (final item in mangas) {
      final nameEl = item.selectFirst("a.manga-title");
      final imageEl = item.selectFirst("img.manga-cover");
      final urlEl = item.selectFirst("a.manga-link");
      if (nameEl == null || urlEl == null) continue;
      MManga manga = MManga();
      manga.name = nameEl.text.trim();
      manga.imageUrl = imageEl?.attr("src") ?? "";
      manga.link = urlEl.attr("href");
      if (!manga.link!.startsWith("http")) {
        manga.link = "${source.baseUrl}${manga.link}";
      }
      mangaList.add(manga);
    }
    // Pagination: check if next page exists (simple heuristic)
    bool hasNextPage = parseHtml(html).selectFirst("a.next-page") != null;
    return MPages(mangaList, hasNextPage);
  }

  @override
  Future<MPages> search(String query, int page, FilterList filterList) async {
    final filters = filterList.filters;
    String url = "${source.baseUrl}/search?q=${Uri.encodeComponent(query)}&page=$page";

    // Apply filters to query string
    for (var filter in filters) {
      if (filter.type == "GenreFilter") {
        for (var item in filter.state) {
          if (item.state == 1) {
            url += "&genres=${Uri.encodeComponent(item.value)}";
          } else if (item.state == 2) {
            url += "&exclude_genres=${Uri.encodeComponent(item.value)}";
          }
        }
      } else if (filter.type == "StatusFilter") {
        url += "&status=${filter.values[filter.state].value}";
      } else if (filter.type == "SortFilter") {
        url += "&sort=${filter.values[filter.state].value}";
      } else if (filter.type == "TypeFilter") {
        url += "&type=${filter.values[filter.state].value}";
      }
    }

    final res = (await client.get(
      Uri.parse(url),
      headers: getHeader(source.baseUrl),
    )).body;
    return parseMangaList(res);
  }

  @override
  Future<MManga> getDetail(String url) async {
    final statusList = [
      {"Ongoing": 0, "Completed": 1, "Hiatus": 2, "Cancelled": 3}
    ];
    final res = (await client.get(
      Uri.parse(url),
      headers: getHeader(source.baseUrl),
    )).body;
    final doc = parseHtml(res);

    MManga manga = MManga();
    manga.author = doc.selectFirst("span.author")?.text ?? "Unknown";
    manga.description = doc.selectFirst("div.description")?.text ?? "";
    final status = doc.selectFirst("span.status")?.text ?? "Unknown";
    manga.status = parseStatus(status, statusList);
    manga.genre = doc.select("a.genre").map((e) => e.text.trim()).toList();

    // Chapter list
    List<MChapter> chapters = [];
    final chapterItems = doc.select("ul.chapter-list > li");
    for (final item in chapterItems) {
      MChapter chapter = MChapter();
      final link = item.selectFirst("a");
      final dateEl = item.selectFirst("span.chapter-date");
      if (link == null) continue;
      chapter.name = link.text.trim();
      chapter.url = link.attr("href");
      if (!chapter.url!.startsWith("http")) {
        chapter.url = "${source.baseUrl}${chapter.url}";
      }
      chapter.dateUpload = dateEl?.attr("data-time") ?? dateEl?.text;
      chapter.scanlator = item.selectFirst("span.scanlator")?.text ?? "";
      chapters.add(chapter);
    }
    manga.chapters = chapters;
    return manga;
  }

  @override
  Future<List<String>> getPageList(String url) async {
    final res = (await client.get(
      Uri.parse(url),
      headers: {
        "Referer": source.baseUrl,
        "Cookie": "adult=true",
      },
    )).body;
    final doc = parseHtml(res);
    // Extract image URLs from the reader
    final images = doc.select("div.reader-container img");
    if (images.isNotEmpty) {
      return images.map((img) => img.attr("src")).where((s) => s.isNotEmpty).toList();
    }
    // Fallback: look for a script containing image array
    final script = doc.select("script").firstWhere(
      (s) => s.text.contains("var images ="),
      orElse: () => null as MElement?,
    );
    if (script != null) {
      final text = script.text;
      final start = text.indexOf("var images = [") + 13;
      final end = text.indexOf("];", start);
      if (start != -1 && end != -1) {
        final jsonStr = "[${text.substring(start, end)}]";
        final List<dynamic> urls = jsonDecode(jsonStr);
        return urls.map((u) => u.toString()).toList();
      }
    }
    return [];
  }

  @override
  List<dynamic> getFilterList() {
    return [
      TextFilter("SearchFilter", "Search..."), // not used but placeholder
      SelectFilter("SortFilter", "Sort By", 0, [
        SelectFilterOption("Latest Update", "latest"),
        SelectFilterOption("Popular", "popular"),
        SelectFilterOption("Title A-Z", "title_asc"),
        SelectFilterOption("Title Z-A", "title_desc"),
      ]),
      SelectFilter("StatusFilter", "Status", 0, [
        SelectFilterOption("All", ""),
        SelectFilterOption("Ongoing", "ongoing"),
        SelectFilterOption("Completed", "completed"),
        SelectFilterOption("Hiatus", "hiatus"),
        SelectFilterOption("Cancelled", "cancelled"),
      ]),
      SelectFilter("TypeFilter", "Type", 0, [
        SelectFilterOption("All", ""),
        SelectFilterOption("Manga", "manga"),
        SelectFilterOption("Manhwa", "manhwa"),
        SelectFilterOption("Manhua", "manhua"),
        SelectFilterOption("Comic", "comic"),
      ]),
      GroupFilter("GenreFilter", "Genres", [
        TriStateFilter("Action", "action"),
        TriStateFilter("Adventure", "adventure"),
        TriStateFilter("Comedy", "comedy"),
        TriStateFilter("Drama", "drama"),
        TriStateFilter("Fantasy", "fantasy"),
        TriStateFilter("Horror", "horror"),
        TriStateFilter("Romance", "romance"),
        TriStateFilter("Sci-Fi", "sci-fi"),
        TriStateFilter("Slice of Life", "slice_of_life"),
        TriStateFilter("Supernatural", "supernatural"),
        TriStateFilter("Mystery", "mystery"),
        TriStateFilter("Psychological", "psychological"),
        TriStateFilter("Tragedy", "tragedy"),
        TriStateFilter("Isekai", "isekai"),
        TriStateFilter("Martial Arts", "martial_arts"),
        TriStateFilter("Harem", "harem"),
        TriStateFilter("Ecchi", "ecchi"),
        TriStateFilter("Mature", "mature"),
        TriStateFilter("Shounen", "shounen"),
        TriStateFilter("Shoujo", "shoujo"),
        TriStateFilter("Seinen", "seinen"),
        TriStateFilter("Josei", "josei"),
      ]),
    ];
  }
}

Map<String, String> getHeader(String url) {
  return {
    "Referer": "$url/",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
  };
}

WeebCentral main(MSource source) {
  return WeebCentral(source: source);
}